### PR TITLE
feat(geojson): GeometryCollection with Generics

### PR DESCRIPTION
**Pull request by wangxingkai**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/geojson/index.d.ts [hunk 1]
```diff
@@ -129,27 +129,27 @@ export interface Polygon extends GeoJsonObject {
...
 /**
  * Geometry Collection
  * https://tools.ietf.org/html/rfc7946#section-3.1.8
  */
-export interface GeometryCollection extends GeoJsonObject {
-    type: "GeometryCollection";
-    geometries: Geometry[];
+export interface GeometryCollection<G extends Geometry | null = Geometry> extends GeoJsonObject {
+    type: 'GeometryCollection';
+    geometries: G[];
```
Comment (wangxingkai) [inline types/geojson/index.d.ts:142]: this is the real change in this PR.

my VSCode fixes some inconsistent code styles.

Comment (danvk) [reply types/geojson/index.d.ts:142]: While a Feature's `geometry` may be set to `null` ([ref][1]), I don't see anything in the spec that indicates that `null` is permitted as an element in a `GeometryCollection` ([ref][2]). Please disallow this.

The RFC also says that "implementations SHOULD avoid nested GeometryCollections" which seems like something that could be enforced here and tested.

[1]: https://datatracker.ietf.org/doc/html/rfc7946#section-3.2
[2]: https://datatracker.ietf.org/doc/html/rfc7946#section-3.1.8

**Review by danvk — CHANGES_REQUESTED** [review]

File: types/geojson/index.d.ts (hunk 1 above)
Comment (wangxingkai) [reply types/geojson/index.d.ts]: oh, yes. that is a mistake. had it fixed in the new commit.

I also noticed the two "SHOULD" in the RFC. as the linked RFC ["Key words for use in RFCs to Indicate Requirement Levels"](https://datatracker.ietf.org/doc/html/rfc2119#section-3) explains that "SHOULD" means "there may exist valid reasons in particular circumstances". so I thought we probably better allow it?

Comment (danvk) [reply types/geojson/index.d.ts]: Fair enough. Could you add a test that `null` is invalid in a `GeometryCollection` and another that allows nested `GeometryCollections` with a comment saying exactly what you said above.

Otherwise looks good to me.

**Review by danvk — APPROVED** [review]
See comment above, otherwise approved.

**Review by icholy — CHANGES_REQUESTED** [review]
~Please revert all unrelated formatting changes, or at least keep everything consistent.~

edit: use double quotes instead of single quotes. That seems to be the convention in this repo.

**Comment by wangxingkai** [comment]
> ~Please revert all unrelated formatting changes, or at least keep everything consistent.~
> 
> edit: use double quotes instead of single quotes. That seems to be the convention in this repo.

Have changed to double quote.

btw, we have other existing single quote usages here:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/3de6548987506fcab4cea60065338f5f4ab9d934/types/geojson/index.d.ts#L20
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/3de6548987506fcab4cea60065338f5f4ab9d934/types/geojson/index.d.ts#L26

---

This repo is actually using single quote. My VSCode was just respecting the repo's setting.

https://github.com/DefinitelyTyped/DefinitelyTyped#can-i-changeenforce-formatting-settings-for-modules

It mentioned that [.editorconfig](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/.editorconfig) and [.prettierrc.json](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/.prettierrc.json) are used for keeping the formatting consistent.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/21c9aaddaac25b74f72f3003454af8fed4033052/.prettierrc.json#L4


https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes

> Prettier is set up on this repo, so you can run `npm run prettier -- --write path/to/package/**/*.ts`.

In this section, it mentioned to use this this script to format the code for each package.

so if we run `npm run prettier -- --write "types/geojson/**/*.ts"`, we can see that there are quite a bit inconsistencies in `geojson` package.

**Review by icholy — APPROVED** [review]

**Review by navya9singh — APPROVED** [review]

**Comment by wangxingkai** [comment]
Ready to merge

**Comment by icholy** [comment]
@wangxingkai cloud use a follow up PR to fix the style to follow the prettier config.

**Comment by rorrobg** [comment]
Hi, this PR cause the next Issue: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/61254

**"@types/geojson - error TS2456: Type alias 'Geometry' circularly references itself"**

Apparently it's caused by the line `export interface GeometryCollection<G extends Geometry = Geometry> extends GeoJsonObject {`

I use typescript v3.5.3.

**Comment by rorrobg** [comment]
@wangxingkai

**Comment by icholy** [comment]
@rorrobg either update TS to 3.7+ or don't update to the latest version of the geojson types.
